feat: add warning when create/update connectin API secret on S3 type but no AWS_S3_BUCKET or with "" as value

zdtsw opened the pull request:
when a new secret is created or  any secret is updated + it has `opendatahub.io/connection-type-protocol: s3` or `opendatahub.io/connection-type-ref: s3` as annotation
- if the secret does not have AWS_S3_BUCKET we show a warning but still allowed 
- if the secret has AWS_S3_BUCKET but has an emtpy string , we show a warning but still allowed
- create a VAP and a VAPB since the standard webhook does not have "warning" only "allow/deny/error" and it is not clear if only show info log in the operator pod

<!--- 
Many thanks for submitting your Pull Request ❤️!

Please complete the following sections for a smooth review.
-->

## Description
<!--- Describe your changes in detail -->

<!--- Link your JIRA and related links here for reference. -->

https://redhat-internal.slack.com/archives/C08T1K4FQVC/p1760112449206949

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
new local build: quay.io/wenzhou/opendatahub-operator-catalog:v3.2.2732-1105

## Newest test from 7th Nov
quay.io/wenzhou/opendatahub-operator-catalog:v3.2.2732-1107
tested on OCP 4.20
create DSCI and DSC with kserve Managed
create secret
```
apiVersion: v1
kind: Secret
metadata:
  name: test-bucket-empty-value-1107
  annotations:
    opendatahub.io/connection-type-protocol: s3
data:
  AWS_ACCESS_KEY_ID: MQ==
  AWS_DEFAULT_REGION: NA==
  AWS_S3_BUCKET: ''
  AWS_S3_ENDPOINT: Mw==
  AWS_SECRET_ACCESS_KEY: Mg==
type: Opaque
```
warning
<img width="773" height="250" alt="Screenshot From 2025-11-07 08-32-26" src="https://github.com/user-attachments/assets/e7d13c55-e442-4dd5-bfdc-f80d7e2e1e60" />


## New test
use old annotation and not has AWS_S3_BUCKET 
<img width="711" height="305" alt="Screenshot From 2025-11-05 10-15-17" src="https://github.com/user-attachments/assets/98067605-c1b7-484d-947d-3ecd04eb3698" />
use new annotation and not has AWS_S3_BUCKET 
<img width="711" height="305" alt="Screenshot From 2025-11-05 10-14-26" src="https://github.com/user-attachments/assets/402e7751-1bb1-4d0d-94d0-03ac7115d14f" />
use old annotation and has AWS_S3_BUCKET but no value set
<img width="711" height="305" alt="Screenshot From 2025-11-05 10-13-46" src="https://github.com/user-attachments/assets/565083fc-a45a-4265-bf34-44e952c876be" />

## OLD test
<img width="635" height="179" alt="Screenshot From 2025-10-21 15-30-57" src="https://github.com/user-attachments/assets/cb7d12e5-81c3-4aa8-865c-f1bb6d33b258" />


## Screenshot or short clip
<!--- If applicable, attach a screenshot or a short clip demonstrating the feature. -->

## Merge criteria
<!--- This PR will be merged by any repository approver when it meets all the points in the checklist -->
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->

- [ ] You have read the [contributors guide](https://github.com/opendatahub-io/opendatahub-operator/blob/incubation/CONTRIBUTING.md).
- [ ] Commit messages are meaningful - have a clear and concise summary and detailed explanation of what was changed and why.
- [ ] Pull Request contains a description of the solution, a link to the JIRA issue, and to any dependent or related Pull Request.
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [ ] The developer has manually tested the changes and verified that the changes work
- [ ] The developer has run the integration test pipeline and verified that it passed successfully

### E2E test suite update requirement

When bringing new changes to the operator code, such changes are by default required to be accompanied by extending and/or updating the E2E test suite accordingly.

To opt-out of this requirement:
1. **Please inspect the [opt-out guidelines](https://github.com/opendatahub-io/opendatahub-operator/blob/main/docs/e2e-update-requirement-guidelines.md)**, to determine if the nature of the PR changes allows for skipping this requirement
2. If opt-out is applicable, provide justification in the dedicated `E2E update requirement opt-out justification` section below
3. Check the checkbox below:
- [ ] Skip requirement to update E2E test suite for this PR
4. Submit/save these changes to the PR description. This will automatically trigger the check.

#### E2E update requirement opt-out justification
<!--- If you checked the box above, please provide a short summary of reasons for opting-out of this requirement -->
<!--- This section can be left empty if you're not opting out of the E2E requirement -->


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Connection API manifests now deploy with the platform so Connection API resources are applied automatically.
  * Adds a validating admission policy for S3 connection secrets that warns when AWS_S3_BUCKET is missing/empty for S3-annotated secrets and can reject invalid creations/updates.

* **Tests**
  * New e2e checks verify the S3 validation policy and its binding are created and active when the component is enabled.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

zdtsw commented:
> The screenshot you show, is it from OpenShift console?

yes, if create resource(screct) from console, it has this warning.
if create from "oc apply" locally, will get warning as well.

carlkyrillos commented on config/connectionAPI/secret_s3_validation.yaml:
```diff
@@ -0,0 +1,37 @@
+      - apiGroups: [""]
+        apiVersions: ["v1"]
+        operations: ["CREATE", "UPDATE"]
+        resources: ["secrets"]
```
@zdtsw It looks like this VAP is going to run when any secret across the entire cluster is created or updated which may impact performance. Can we use `.spec.matchConstraints.objectSelector.matchLabels` to narrow the scope?

zdtsw replied:
performance wise i guess it is ok, it wont check any existing ones, and only on create, update. similar to our webhook on certain type

carlkyrillos replied:
Won't the s3 connection secret have either the deprecated `opendatahub.io/connection-type-ref` annotation or the new `opendatahub.io/connection-type-protocol` annotation?

zdtsw replied:
removed

zdtsw commented:
> @zdtsw I verified the changes and everything is working as expected except I'm not seeing the warning if I create an s3 `-protocol` secret with an empty value for `AWS_S3_BUCKET`

this is my test
<img width="773" height="752" alt="Screenshot From 2025-11-07 08-15-49" src="https://github.com/user-attachments/assets/538b3955-80fa-4e97-9627-3158f0cd8e3e" />


<img width="1626" height="728" alt="Screenshot From 2025-11-07 08-15-36" src="https://github.com/user-attachments/assets/bf800b83-7192-4284-8478-fd21cd001c4d" />

zdtsw commented:
> @zdtsw Minor issue but the warning is currently rendered as this:
> 
> ```
> Secret s3-type-protocol violates policy 299 - "Validation failed for ValidatingAdmissionPolicy 'connectionapi-check-s3-bucket' with binding 'connectionapi-check-s3-bucket-binding': Secret \"s3-type-protocol\" has S3 connection type annotation but is missing \"AWS_S3_BUCKET\" data key or with empty string as value"
> ```
> 
> It would be great if we can fix the formatting so it doesn't show the `\` in front of the `"`.


updated and created new build, updated PR description with test result.

carlkyrillos commented:
/lgtm
